R-package/R/lgb.Booster.R

```diff
@@ -727,6 +728,20 @@ Booster <- R6::R6Class(
 
     },
 
+    get_loaded_param = function(handle) {
+      params_str <- .Call(
+        LGBM_BoosterGetLoadedParam_R
+        , handle
+      )
+      params <- jsonlite::fromJSON(params_str)
+      if ("interaction_constraints" %in% names(params)) {
+        params[["interaction_constraints"]] <- lapply(params[["interaction_constraints"]], function(x) x + 1L)
+      }
```

Comment:
Given that this was only necessary for interaction constraints I'm afraid categorical feature's indices aren't being adjusted. Can that be the case?

Reply:
It happens on dataset:
https://github.com/microsoft/LightGBM/blob/2b8fe8b4bdc00a2611442fdee4c45316f08b1c4b/R-package/R/lgb.Dataset.R#L159
https://github.com/microsoft/LightGBM/blob/2b8fe8b4bdc00a2611442fdee4c45316f08b1c4b/R-package/R/lgb.Dataset.R#L184
We may want to work on this later because in this case the saved model's categorical_feature is wrong.

Reply:
As I mentioned in https://github.com/microsoft/LightGBM/pull/4970#pullrequestreview-937407074

1. there are many places where LightGBM's C++ code expects 0-based indexes
2. because R is 1-based, there have been some places where the R package has translated from 1-based inputs to 0-based values
3. this has been done very inconsistently
4. this requires careful thought and a comprehensive plan

For example.... should 0-based indices be converted back to 1-based in `Booster$save_model()`? Or is it expected that the model file produced by R code could be directly read by Python code?

What about in `params`? Is it expected that you could save a dictionary from Python code to a file like `params.json`, then read it in R and immediately pass it to `lgb.train(..., params)`?

It's not obvious to me which choice in these and other situations is "right" or "wrong", and I think that deserves a comprehensive plan with guidelines for when users should expect to see 0-based vs. 1-based values. I'd like to do that work at some point, but it is really difficult and a large amount of effort right.

For the purpose of this PR, I recommend just focusing on preserving the current behavior.

Reply:
I was thinking that maybe the CLI would get the wrong predictions but it doesn't seem like it. IMO I would expect the saved model to be the same regardless of whether it was trained on R or Python, so converting to 0-based indices when saving makes sense to me.

> For the purpose of this PR, I recommend just focusing on preserving the current behavior.

Yes, I don't intend to make changes on this PR, just wanted to highlight that I found it odd.